### PR TITLE
Blog post for 0.16.0

### PR DESCRIPTION
**Pull request by maks-rafalko**
I've already created a `016.` branch in `infection/site`, and will create the same for `infection/infection` soon.

As we already discussed, I need some help with this post to explain what performance improvements we did (and everything else that I missed)

### CODE REVIEW

File: src/_posts/whats-new-in-0.16.0.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+Infection's first step is executing your project's test suite. If it fails, further mutation process is useless, because any Mutant can be marked as Killed just because of the failing tests for the original code.
+
+That's why passing test suite is essential.
+
+If you have already run the test suite to ensure it is passing, and you are providing coverage using `--coverage` then you can use `--skip-initial-test-suite` option to increase performance by not running the entire test suite again.
+
+<p class="tip">Note that it is extremely important that you ensure the test suite is passing when using this option, otherwise test failures would appear like caught mutations and those mutations may be reported incorrectly.</p>
+
+### Other performance improvements
```
Comment (maks-rafalko) [inline src/_posts/whats-new-in-0.16.0.md]: @sanmai I left all the performance-related tickets that were done recently, if you will be able to help, here they are

File: src/_posts/whats-new-in-0.16.0.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+TBD
+
+* https://github.com/infection/infection/pull/1177
+* https://github.com/infection/infection/pull/1172
+* https://github.com/infection/infection/pull/1106
+* https://github.com/infection/infection/pull/1082
+
+## Splitting Infection by separate Packages
+
+Infection has been split by different packages. For example, we have extracted `PhpSpec` and `Codeception` adapter out of the core.
```
Comment (localheinz) [inline src/_posts/whats-new-in-0.16.0.md]:
```suggestion
Infection has been split into different packages. For example, we have extracted `PhpSpec` and `Codeception` adapter out of the core.
```

File: src/_posts/whats-new-in-0.16.0.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+You can read an very detailed explanation about why was it needed in [this RFC](https://github.com/infection/infection/issues/922).
```
Comment (localheinz) [inline src/_posts/whats-new-in-0.16.0.md]:
```suggestion
You can read a very detailed explanation about why was it needed in [this RFC](https://github.com/infection/infection/issues/922).
```

File: src/_posts/whats-new-in-0.16.0.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+Don't worry, if you are using Codeception or PhpSpec, these packages will be automatically installed at the first Infection execution after upgrading to 0.16.0.
+
+List of new packages:
+
+* [`infection/codeception-adapter`](https://github.com/infection/codeception-adapter)
+* [`infection/phpspec-adapter`](https://github.com/infection/phpspec-adapter)
+* [`infection/abstract-testframework-adapter`](https://github.com/infection/abstract-testframework-adapter) - interfaces for any testframework adapter
+* [`infection/extension-installer`](https://github.com/infection/extension-installer) - autodiscovery of installed adapters. No need to add them to `infection.json`, just install a Composer package and you are done. Inspired by PHPStan and Psalm plugin system.
+* [`infection/include-interceptor`](https://github.com/infection/include-interceptor) - Stream Wrapper. Allows to replace included (autoloaded) file with another one (magic that replaces your code with a mutated one)
```
Comment (localheinz) [inline src/_posts/whats-new-in-0.16.0.md]: What do you think about keeping this list sorted by package names?

File: src/_posts/whats-new-in-0.16.0.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+We had this feature [before](/guide/mutation-badge.html), but it worked only for TravisCI.
+
+Now, thanks to [`ondram/ci-detector`](https://github.com/OndraM/ci-detector), Mutation Badge can be used for [all the supported](https://github.com/OndraM/ci-detector#supported-continuous-integration-servers) by this lib CI servers, including GitHub Actions, CircleCI, GitLab and so on.
+
+Feel free to request an additional integration there if the list of currently supported continuous-integration servers does not contain a service you require.
+
+> Read more about how to setup [Mutation Badge](/guide/mutation-badge.html)
+
+## Global ignore feature for Mutators and Profiles
```
Comment (maks-rafalko) [inline src/_posts/whats-new-in-0.16.0.md]: @theofidry feel free to add anything you think should be in this post as well, as there were tons of PRs from you, unfortunately without `0.16.0` milestone. I've added those that found there

File: src/_posts/whats-new-in-0.16.0.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+* [`infection/abstract-testframework-adapter`](https://github.com/infection/abstract-testframework-adapter) - interfaces for any testframework adapter
+* [`infection/extension-installer`](https://github.com/infection/extension-installer) - autodiscovery of installed adapters. No need to add them to `infection.json`, just install a Composer package and you are done. Inspired by PHPStan and Psalm plugin system.
+* [`infection/include-interceptor`](https://github.com/infection/include-interceptor) - Stream Wrapper. Allows to replace included (autoloaded) file with another one (magic that replaces your code with a mutated one) 
+
+
+## Mutation Badge
+
+We had this feature [before](/guide/mutation-badge.html), but it worked only for TravisCI.
+
+Now, thanks to [`ondram/ci-detector`](https://github.com/OndraM/ci-detector), Mutation Badge can be used for [all the supported](https://github.com/OndraM/ci-detector#supported-continuous-integration-servers) by this lib CI servers, including GitHub Actions, CircleCI, GitLab and so on.
```
Comment (localheinz) [inline src/_posts/whats-new-in-0.16.0.md]:
```suggestion
Now, thanks to [`ondram/ci-detector`](https://github.com/OndraM/ci-detector), the mutation badge can be used for [CI servers supported by this library](https://github.com/OndraM/ci-detector#supported-continuous-integration-servers), including GitHub Actions, CircleCI, GitLab and so on.
```

File: src/_posts/whats-new-in-0.16.0.md (hunk 4 above)
Comment (maks-rafalko) [reply src/_posts/whats-new-in-0.16.0.md]: thought about it, but they are groupd "by meaning" right now, which IMO better here

File: src/_posts/whats-new-in-0.16.0.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+It's possible to [use Infection with CI server](/guide/using-with-ci.html), adding `--min-msi` and/or `--min-covered-msi` options, so when the actual results of these metrics are below the threshold, the build fails.
+
+But, when you constantly improve the quality of your tests, MSI metric eventually increases its value (which is good), and now Infection will remind you to update the limits.
+
+```bash
+! [NOTE] The MSI is 2.5787422095418% percent points over the required MSI.     
+!        Consider increasing the required MSI percentage the next time you run 
+!        infection.    
+```
```
Comment (theofidry) [inline src/_posts/whats-new-in-0.16.0.md]:
```suggestion
While this is not the recommended way*, it is possible to use the options `--min-msi` and/or `--min-covered-msi` options to force a certain threshold (the Infection process will exit with the error code `1` if one of those is not reached). When using Infection with those, Infection would also now display a recommendation to increase this threshold when the score is above the configured threshold:

```bash
! [NOTE] The MSI is 2.5787422095418% percent points over the required MSI.     
!        Consider increasing the required MSI percentage the next time you run 
!        infection.    
```

*: The MSI score is but a metric like another. The interesting part is not necessarily the score but which mutations escaped and whether or not those should be caught. Moreover if you are using Infection in an incremental way, i.e. running it only the changed files, this metric will drastically vary from a change to another depending of how testing this part is making it unsuitable for this use case.
```

File: src/_posts/whats-new-in-0.16.0.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+<p class="tip">Note that it is extremely important that you ensure the test suite is passing when using this option, otherwise test failures would appear like caught mutations and those mutations may be reported incorrectly.</p>
+
+### Other performance improvements
+
+TBD
+
+* https://github.com/infection/infection/pull/1177
+* https://github.com/infection/infection/pull/1172
+* https://github.com/infection/infection/pull/1106
+* https://github.com/infection/infection/pull/1082
```
[the remark this answers is not in the document]
Comment (maks-rafalko) [reply src/_posts/whats-new-in-0.16.0.md]: Do you want to add a few words to explain what is going on there?

[the remark this answers is not in the document]
Comment (theofidry) [reply src/_posts/whats-new-in-0.16.0.md]: I'm wondering if we should actually list those PRs in the first place? We detailed what it was about above

[the remark this answers is not in the document]
Comment (maks-rafalko) [reply src/_posts/whats-new-in-0.16.0.md]: I guess we shoudn't. I've added them for @sanmai to remind what was the improvements

File: src/_posts/whats-new-in-0.16.0.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+---
+
+Release: https://github.com/infection/infection/releases/tag/0.16.0
+
+## BC Breaks
+
+* Infection now requires PHP 7.3.12+. If you can't upgrade for some reason upgrade, don't worry, you can still use previous versions.
+* We had [a bug](https://github.com/infection/infection/pull/1105) with the config file loading order. Instead of trying to load the file `infection.json` and if not found, try to load `infection.json.dist`, it was doing this in a reverse order, stopping at `infection.json.dist`. If you have both `infection.json` and `infection.json.dist`, please note that the first will be used like it should, unlike before.
+
+## Performance
```
Comment (theofidry) [inline src/_posts/whats-new-in-0.16.0.md]:
```suggestion
## Performance improvements
```

**Comment by maks-rafalko** [comment]
Thank you all for helping with the blog post

**Pull request merged**